Automattic/jetpack
# Reorg: Iterate on mirror repo publishing

#### Changes proposed in this Pull Request:

Summary:
* DRY in the mirror repo publishing script.
* Rework output in the script for better use of GH Actions.
* Read mirror repo name from composer.json instead of special casing the Jetpack plugin in the script.
* Rework Jetpack's .gitignore and .svnignore stuff to use .gitattributes instead, and handle it in the repo publishing script. Unfortunately `git archive` turned out not to work for the purpose.
* Move build-asset-cdn-json.php into Jetpack's build process.
* Allow the script to publish to non-master mirror branches too.

----

For .gitignore, .svnignore, and .gitattributes, we have four classes of files:
    
1. Files included in both the monorepo and production. We generally ignore this case, treating it as the default when a file doesn't fall into any of the other cases.
2. Files excluded from both the monorepo and production. These remain in .gitignore; they're now the only ones in .gitignore.
3. Files excluded from the monorepo but included in production. These used to be in the bottom section of .gitignore; they've been moved to .gitattributes with the tag "production-include".
4. Files included in the monorepo but excluded from production. These used to be in .svnignore; they've been moved to .gitattributes with the tag "production-exclude".

I also took the opportunity to clean up the three classes.

* We no longer need to list individual subdirectories of vendor in .gitignore, as the new mechanism doesn't get confused.
* Jetpack Autoloader v1 files have been removed from case 3, as they're no longer present.
* Case 4 sees a lot of removals:
  * Git ignores `.git` by default.
  * Several files were left behind in the repo root: `.codeclimate.yml`, `.editorconfig`, `.eslines.json`, `.eslintcache`, `.eslintignore`, `.github`, `.npmrc`, `.nvmrc`, `dangerfile.js`, `docker`, `docs`, `gulpfile.js`, `logs`, `packages`, `postcss.config.js`, `to-test.md`, `webpack.config*`.
  * A bunch of files in `bin/` are now blanket-ignored in `tools/`.
  * A single entry for `.eslintrc.js` covers it everywhere.
  * Several entries don't seem to match anything anymore, or at least anything that isn't already ignored: `**/__snapshots__`, `extensions/**/*.css`, `extensions/**/*.gif`, `extensions/**/*.jpeg`, `extensions/**/*.sass`, `extensions/**/*.snap`, `vendor/automattic/**/.gitattributes`, `vendor/automattic/**/phpunit.xml`, `vendor/automattic/**/phpunit.xml.dist`, `vendor/automattic/**/tests/`,
  * A few others are covered by .gitignore and so don't need re-listing: `vendor/bin`, `vendor/dealerdirect`, `vendor/phpcompatibility`, `vendor/sirbrillig`, `vendor/squizlabs`, `vendor/wp-coding-standards`, `yarn-error.log`,
* Case 4 also has some additions:
  * `.size-limit.js` and `.wp-env.json` were accidentally skipped by the previous scripts, and now need explicit listing.
  * `extensions/**/test` and `modules/**/test` seem intended to be ignored, but had been covered by ignoring every file extension that might appear in them.
  * `modules/masterbar/admin-color-schemes/colors/**/*.scss` seems to have been recently added, and no matching pattern was added to .svnignore yet.

The end result (considering added and removed files) of the generated Automattic/jetpack-production compared to the current SVN looks like
```
!       .svnignore
!       CODE-OF-CONDUCT.md
!       _inc/blocks/vendors~map/mapbox-gl.c47a73a369b133a00458.css
!       _inc/blocks/vendors~map/mapbox-gl.c47a73a369b133a00458.js
!       _inc/blocks/vendors~map/mapbox-gl.c47a73a369b133a00458.rtl.css
?       _inc/blocks/vendors~map/mapbox-gl.ebbf21e4c672155f3a76.css
?       _inc/blocks/vendors~map/mapbox-gl.ebbf21e4c672155f3a76.js
?       _inc/blocks/vendors~map/mapbox-gl.ebbf21e4c672155f3a76.rtl.css
!       _inc/blocks/vendors~swiper.6c77cf0f0df648fd416c.css
!       _inc/blocks/vendors~swiper.6c77cf0f0df648fd416c.js
!       _inc/blocks/vendors~swiper.6c77cf0f0df648fd416c.rtl.css
?       _inc/blocks/vendors~swiper.97fb1069b26372006ed7.css
?       _inc/blocks/vendors~swiper.97fb1069b26372006ed7.js
?       _inc/blocks/vendors~swiper.97fb1069b26372006ed7.rtl.css
!       _inc/build/instant-search/jp-search.chunk-1-f4300f48331e2183fdd9.js
?       _inc/build/instant-search/jp-search.chunk-1-f51ed50616aca95016ab.js
?       _inc/lib/core-api/wpcom-fields/class-wpcom-rest-api-v2-attachment-videopress-data.php
!       bin
!       bin/build-asset-cdn-json.php
!       bin/partner-cancel.sh
!       bin/partner-provision.sh
!       modules/.eslintrc.js
```

#### Jetpack product discussion


1199714989172614-as-1199732467231541

#### Does this pull request change what data or activity we track or use?


No.

#### Testing instructions:






* You should be able to run `GITHUB_ACTOR=... GITHUB_REF=refs/heads/master .github/files/update-project-mirrors.sh` locally and then verify the sanity of the repos generated. It won't try to push to GitHub as long as you don't have `CI` set in your environment.
* Post-merge, it should start pushing to a feature/reorg branch on all the mirror repos as things are pushed to the feature/reorg branch in the monorepo.

#### Proposed changelog entry for your changes:


* None needed, will be subsumed into feature/reorg.
